shard_connection.cpp

```diff
@@ -542,6 +572,23 @@ void shard_connection::send_set_command(struct timeval* sent_time, const char *k
     push_req(new request(rt_set, cmd_size, sent_time, 1));
 }
 
+
+/*
+ * empty rate limit command:
+ *
+ * we send the empty rate limit command so that when a client has exceed is quota
+ * instead of writing new requests, it will just submit empty event to the event loop.
+ * This is needed since the normal rate limit based on suspending the thread that is
+ * exceeding the quota can affect the other clients on that same thread and the
+ * latency of the request as well.
+ * In this way we are not effecting other clients overall latency measurement.
+ */
+void shard_connection::send_empty_rate_limit_event() {
+    if ( bufferevent_write(m_bev,"^\r\n",3) == 0 ){
```

Comment:
Please check it again... when we use bufferevent_write we are adding data to the output buffer of m_bev ... means.. writing data on the socket to the server

Reply:
you're totally right @yaacovhazan-Redislabs . 
confirmed that yes, we're sending the "rate-limiting-dummy" command to redis.  will correct it 👍

Reply:
@filipecosta90 The libevent `bufferevent` has some built in rate limiting mechanism. It's obviously based on throughput and knows nothing about requests/responses but perhaps it's something you can leverage for this. I don't remember how it works or what APIs are exposed so it may or may not be possible but worth looking.

Reply:
hi there @yossigo and @yaacovhazan-Redislabs, sorry for the delayed reply. Considering that the histogram and the rate-limiting features are separate ones and one is stalling other I would go for splitting this PR into three:
- 1st one ( this one ) -  with overall percentile analysis feature
- 2nd one - extended percentile over time ( as discussed with @yossigo on January )
- 3rd one - rate limiting feature using libevent's Bufferevents and Rate-limiting features 
In that manner we would not delay the possibility of having extended metrics on memtier. What do you guys think?